src/app/hooks/use-theme.ts

```diff
@@ -32,5 +32,11 @@ export const useTheme = () => {
 
 	const resetTheme = () => setTheme("system");
 
-	return { isDarkMode, resetTheme, setProfileTheme, setTheme, theme };
+	const resetProfileTheme = (profile: Contracts.IProfile) => {
```

Comment:
It is a good to wrap this and previous functions in useCallback to stop optional rerenders when passing one of this functions to children.
```suggestion
	const resetProfileTheme = useCallback((profile: Contracts.IProfile) => {
```

Reply:
i rarely use memoization (`useCallback`, `useMemo`,...), because memoization isn't cheap, It occupies the memory space for storing the memoised values.
memoization is a way to avoid recomputing expensive calculations, but i can't see any expensive calculations in this file

> React doc: ["Is it OK to use arrow functions in render methods?"](https://reactjs.org/docs/faq-functions.html#is-it-ok-to-use-arrow-functions-in-render-methods)

Reply:
If you are returning functions and using them outside the hook it still occupies the memory to save these functions in the scope, not too many changes with it if you are using `useCallback`.
The problem is with the redeclaration of the function of each custom hook execution. That means every execution you have a new functional object and all child hooks, or components, or handlers that have this dependency will be recalculated again. With `useCallback` you all-time return the same function object, that's avoiding rerenders.

> [React Performance: Event Handlers using `useCallback` hook.](https://medium.com/@KTAsim/react-performance-event-handlers-using-usecallback-hook-9e4a06f8bb2f)

Reply:
there is no benefit to using `useCallback` in [this hook](https://github.com/PayvoHQ/wallet/blob/develop/src/app/hooks/use-theme.ts) because it doesn't do any expensive calculations

> [Kent C. Dodds](https://github.com/kentcdodds): "Performance optimizations are not free. They ALWAYS come with a cost but do NOT always come with a benefit to offset that cost." ([from this article](https://kentcdodds.com/blog/usememo-and-usecallback#whats-the-point))

Reply:
No expensive calculations are here, that's right.
But just after the sentences from the artical:

> There are specific reasons both of these hooks are built-into React:
> 1. Referential equality

that's what I'm talking about. Just don't recreate a new handler function each time.

Reply:
don’t forget that `useCallback()` returning the same function object, still, the inline function is re-created on every re-rendering (`useCallback()` just skips it).

[useTheme](https://github.com/PayvoHQ/wallet/blob/fix/reset-appearance-settings/src/app/hooks/use-theme.ts#L7) with `useCallback` should be something like this:

```js
const setTheme = useCallback((theme: Theme) => {
	setThemeSource(theme);

	if (shouldUseDarkColors()) {
		document.body.classList.add("theme-dark");
		document.body.classList.remove("theme-light");
	} else {
		document.body.classList.add("theme-light");
		document.body.classList.remove("theme-dark");
	}
}, []);

const setProfileTheme = useCallback((profile: Contracts.IProfile) => {
	const profileTheme = profile.appearance().get("theme");
	const hasDifferentTheme = shouldUseDarkColors() !== (profileTheme === "dark");

	/* istanbul ignore else */
	if (hasDifferentTheme || !document.body.classList.contains(`theme-${theme}`)) {
		setTheme(profileTheme as Theme);
	}
}, [setTheme, theme]);

const resetTheme = useCallback(() => setTheme("system"), [setTheme]);

const resetProfileTheme = useCallback((profile: Contracts.IProfile) => {
	resetTheme();

	profile.settings().set(Contracts.ProfileSetting.Theme, shouldUseDarkColors() ? "dark" : "light");
}, [resetTheme]);
```

at the first function because there is an empty dependency ([]), `useCallback()` is called every time the component re-renders. even though it returns the same function object, still the inline function is created on every render. ([more details](https://kentcdodds.com/blog/usememo-and-usecallback#why-is-usecallback-worse))

because of this, there is no need to use `useCallback` here

Reply:
I agree with the disadvantages of `useCallback` because of creating an array object besides the function object.
In the cases like this, I try to weigh the pros and cons:
- If it's a function used only in this component when rendering – storing it would be unnecessary.
- If it's a handler that we pass through several components – better an extra array than unnecessary DOM tree recalculation.

Also, would be good to memorize the return value of the hook by the same reasons. If we use it like value anywhere.
```js
return useMemo(() => ({ isDarkMode, resetProfileTheme, resetTheme, setProfileTheme, setTheme, theme }), [
	isDarkMode,
	resetProfileTheme,
	resetTheme,
	setProfileTheme,
	setTheme,
	theme,
]);

```